File: app/src/main/kotlin/com/saqib/googlepay/gpay/jwt/Jwt.java
package com.saqib.googlepay.gpay.jwt;

import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.saqib.googlepay.gpay.Config;

import net.oauth.jsontoken.JsonToken;
import net.oauth.jsontoken.crypto.RsaSHA256Signer;

import org.joda.time.Instant;

import java.security.InvalidKeyException;
import java.security.Key;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;

import static net.oauth.jsontoken.JsonToken.ALGORITHM_HEADER;
import static net.oauth.jsontoken.JsonToken.KEY_ID_HEADER;

public class Jwt {
    private String audience;
    private String type;
    private String iss;
    private ArrayList<String> origins;
    private Instant iat;
    private JsonObject payload;
    private RsaSHA256Signer signer;
    private Config config;

    public Jwt(Context context) {
        config = Config.getInstance(context);
        audience = config.getAudience();
        type = config.getJwtType();
        iss = config.getServiceAccountEmailAddress();
        origins = config.getOrigins();
        iat = new Instant(Calendar.getInstance().getTimeInMillis() - 5000L);
        payload = new JsonObject();

        try {
            this.signer = new RsaSHA256Signer(this.iss, config.KEY_ID, config.getServiceAccountPrivateKey());
        } catch (InvalidKeyException e) {
            e.printStackTrace();
        }
    }

    public void addTransitClass(JsonElement resourcePayload) {
        if (this.payload.get("transitClasses") == null) {
            JsonArray transitClasses = new JsonArray();
            this.payload.add("transitClasses", transitClasses);

        }
        JsonArray newPayload = (JsonArray) this.payload.get("transitClasses");
        newPayload.add(resourcePayload);
        this.payload.add("transitClasses", newPayload);

        return;
    }

    public void addTransitObject(JsonElement resourcePayload) {
        if (payload.get("transitObjects") == null) {
            JsonArray transitObjects = new JsonArray();
            payload.add("transitObjects", transitObjects);
        }
        JsonArray newPayload = (JsonArray) payload.get("transitObjects");
        newPayload.add(resourcePayload);
        payload.add("transitObjects", newPayload);
    }

    private JsonToken generateUnsignedJwt() {
        JsonToken token = new JsonToken(this.signer);
        token.setAudience(this.audience);
        token.setParam("typ", this.type);
        token.setIssuedAt(this.iat);

        Gson gson = new Gson();
        token.getPayloadAsJsonObject().add("payload", this.payload);
        token.getPayloadAsJsonObject().add("origins", gson.toJsonTree(this.origins));
        return token;
    }

    public String generateSignedJwt() {
        JsonToken jwtToSign = this.generateUnsignedJwt();
        return getSignedJwt(jwtToSign);
    }

    private Map<String, Object> getHeaders(JsonToken jwtToSign) {
        Map<String, Object> headers = new HashMap<>();
        headers.put("typ", "JWT");
        headers.put(ALGORITHM_HEADER, jwtToSign.getSignatureAlgorithm().getNameForJson());
        String keyId = jwtToSign.getKeyId();
        if (keyId != null) {
            headers.put(KEY_ID_HEADER, keyId);
        }
        return headers;
    }

    public String getSignedJwt(JsonToken jwtToSign) {
        Key key = config.getServiceAccountPrivateKey();
        return Jwts.builder()
                .setPayload(jwtToSign.getPayloadAsJsonObject().toString())
                .setHeader(getHeaders(jwtToSign))
                .signWith(key, SignatureAlgorithm.RS256)
                .compact();
    }
}

File: app/src/main/kotlin/com/saqib/googlepay/gpay/jwt/Jwt.kt
package com.saqib.googlepay.gpay.jwt

import android.content.Context
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.saqib.googlepay.gpay.GooglePayConfig
import com.saqib.googlepay.gpay.GooglePayConfig.getInstance
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import net.oauth.jsontoken.JsonToken
import net.oauth.jsontoken.crypto.RsaSHA256Signer
import org.joda.time.Instant
import java.security.InvalidKeyException
import java.security.Key
import java.util.*

class Jwt(val context: Context) {
    private val audience: String
    private val type: String
    private val iss: String
    private val origins: ArrayList<String>
    private val iat: Instant
    private val payload: JsonObject
    private var signer: RsaSHA256Signer? = null
    private val config: GooglePayConfig = getInstance(context)

    fun addTransitClass(resourcePayload: JsonElement?) {
        if (payload["transitClasses"] == null) {
            val transitClasses = JsonArray()
            payload.add("transitClasses", transitClasses)
        }
        val newPayload = payload["transitClasses"] as JsonArray
        newPayload.add(resourcePayload)
        payload.add("transitClasses", newPayload)
        return
    }

    fun addTransitObject(resourcePayload: JsonElement?) {
        if (payload["transitObjects"] == null) {
            val transitObjects = JsonArray()
            payload.add("transitObjects", transitObjects)
        }
        val newPayload = payload["transitObjects"] as JsonArray
        newPayload.add(resourcePayload)
        payload.add("transitObjects", newPayload)
    }

    private fun generateUnsignedJwt(): JsonToken {
        val token = JsonToken(signer)
        token.audience = audience
        token.setParam("typ", type)
        token.issuedAt = iat
        val gson = Gson()
        token.payloadAsJsonObject.add("payload", payload)
        token.payloadAsJsonObject.add("origins", gson.toJsonTree(origins))
        return token
    }

    fun generateSignedJwt(): String {
        val jwtToSign = generateUnsignedJwt()
        return getSignedJwt(jwtToSign)
    }

    private fun getHeaders(jwtToSign: JsonToken): Map<String, Any> {
        val headers: MutableMap<String, Any> = HashMap()
        headers["typ"] = "JWT"
        headers[JsonToken.ALGORITHM_HEADER] = jwtToSign.signatureAlgorithm.nameForJson
        val keyId = jwtToSign.keyId
        if (keyId != null) {
            headers[JsonToken.KEY_ID_HEADER] = keyId
        }
        return headers
    }

    fun getSignedJwt(jwtToSign: JsonToken): String {
        val key: Key = config.serviceAccountKey
        return Jwts.builder()
            .setPayload(jwtToSign.payloadAsJsonObject.toString())
            .setHeader(getHeaders(jwtToSign))
            .signWith(key, SignatureAlgorithm.RS256)
            .compact()
    }

    init {
        audience = config.AUDIENCE
        type = config.JWT_TYPE
        iss = config.SERVICE_ACCOUNT_EMAIL_ADDRESS
        origins = config.origins
        iat = Instant(Calendar.getInstance().timeInMillis - 5000L)
        payload = JsonObject()
        try {
            signer = RsaSHA256Signer(iss, config.KEY_ID, config.serviceAccountKey)
        } catch (e: InvalidKeyException) {
            e.printStackTrace()
        }
    }
}
